hudi-client/hudi-client-common/src/test/java/org/apache/hudi/table/action/savepoint/TestSavepointActionExecutor.java

```diff
@@ -0,0 +1,120 @@
+    when(mockTable.getActiveTimeline().readCleanMetadata(cleanInstant)).thenReturn(cleanMetadata);
+
+    SavepointActionExecutor executor = new SavepointActionExecutor(engineContext, null, mockTable, instantTime, "user", "comment");
+    assertEquals(expectedInstant, executor.getLastCommitRetained());
+  }
+
+  @Test
+  void testLastCommitRetained_requestedClean() throws IOException {
+    String expectedInstant = "20240101101012345";
+    String cleanRequested = "20240111101012345";
+    HoodieInstant cleanInstant = new HoodieInstant(HoodieInstant.State.REQUESTED, HoodieTimeline.CLEAN_ACTION, cleanRequested,
+        InstantComparatorV2.REQUESTED_TIME_BASED_COMPARATOR);
+    InstantGenerator instantGenerator = mock(InstantGenerator.class);
+    when(mockTable.getInstantGenerator()).thenReturn(instantGenerator);
+    when(instantGenerator.createNewInstant(HoodieInstant.State.REQUESTED, HoodieTimeline.CLEAN_ACTION, cleanRequested))
+        .thenReturn(cleanInstant);
+
+    when(mockTable.getCompletedCommitsTimeline().firstInstant()).thenReturn(Option.of(new HoodieInstant(HoodieInstant.State.COMPLETED, HoodieTimeline.COMMIT_ACTION, expectedInstant,
+        InstantComparatorV2.REQUESTED_TIME_BASED_COMPARATOR)));
+    when(mockTable.getCleanTimeline().lastInstant()).thenReturn(Option.of(cleanInstant));
+
+    HoodieCleanerPlan cleanerPlan = new HoodieCleanerPlan();
+    HoodieActionInstant lastInstantToRetain = HoodieActionInstant.newBuilder().setAction(HoodieTimeline.COMMIT_ACTION)
+        .setState(HoodieInstant.State.COMPLETED.name()).setTimestamp(expectedInstant).build();
+    cleanerPlan.setEarliestInstantToRetain(lastInstantToRetain);
+    when(mockTable.getActiveTimeline().readCleanerPlan(cleanInstant)).thenReturn(cleanerPlan);
+
+    SavepointActionExecutor executor = new SavepointActionExecutor(engineContext, null, mockTable, instantTime, "user", "comment");
+    assertEquals(expectedInstant, executor.getLastCommitRetained());
+  }
+
+  @Test
+  void testLastCommitRetained_completedCleanWithoutEarliestInstantRetained() throws IOException {
+    String expectedInstant = "20240110101012345";
+    HoodieInstant cleanInstant = new HoodieInstant(HoodieInstant.State.COMPLETED, HoodieTimeline.CLEAN_ACTION, "20240111101012345",
+        InstantComparatorV2.REQUESTED_TIME_BASED_COMPARATOR);
+    when(mockTable.getCompletedCommitsTimeline().firstInstant()).thenReturn(Option.of(new HoodieInstant(HoodieInstant.State.COMPLETED, HoodieTimeline.COMMIT_ACTION, expectedInstant,
+        InstantComparatorV2.REQUESTED_TIME_BASED_COMPARATOR)));
+    when(mockTable.getCleanTimeline().lastInstant()).thenReturn(Option.of(cleanInstant));
+
+    HoodieCleanMetadata cleanMetadata = new HoodieCleanMetadata();
+    when(mockTable.getActiveTimeline().readCleanMetadata(cleanInstant)).thenReturn(cleanMetadata);
+
+    MockHoodieTimeline mockTimeline = new MockHoodieTimeline(Stream.of("20240108101012345", "20240109101012345", expectedInstant), Stream.empty());
```

Comment:
can we add more completed instants greater than latest completed clean in this test as well.

Reply:
Added an instant greater than the clean